lib/kubernetes-deploy/deferred_summary_logging.rb

```diff
@@ -0,0 +1,92 @@
+module KubernetesDeploy
+  # Adds the methods kubernetes-deploy requires to your logger class.
+  # These methods include helpers for logging consistent headings, as well as facilities for
+  # displaying key information later, in a summary section, rather than when it occurred.
+  module DeferredSummaryLogging
+    attr_reader :summary
+    def initialize(*args)
+      reset
+      super
+    end
+
+    def reset
+      @summary = DeferredSummary.new
+      @current_phase = 0
+    end
+
+    def blank_line(level = :info)
+      public_send(level, "")
+    end
+
+    def phase_heading(phase_name)
+      @current_phase += 1
+      heading("Phase #{@current_phase}: #{phase_name}")
+    end
+
+    def heading(text, secondary_msg='', secondary_msg_color=:blue)
+      padding = (100.0 - (text.length + secondary_msg.length)) / 2
+      blank_line
+      part1 = ColorizedString.new("#{'-' * padding.floor}#{text}").blue
+      part2 = ColorizedString.new(secondary_msg).colorize(secondary_msg_color)
+      part3 = ColorizedString.new("#{'-' * padding.ceil}").blue
+      info(part1 + part2 + part3)
+    end
+
+    # Outputs the deferred summary information saved via @logger.summary.add_action and @logger.summary.add_paragraph
+    def print_summary(success)
+      if success
+        heading("Result: ", "SUCCESS", :green)
+        level = :info
+      else
+        heading("Result: ", "FAILURE", :red)
+        level = :fatal
+      end
+
+      public_send(level, summary.actions_sentence)
+      summary.paragraphs.each do |para|
+        blank_line(level)
+        msg_lines = para.split("\n")
+        msg_lines.each { |line| public_send(level, line) }
+      end
+    end
+
+    class DeferredSummary
+      attr_reader :paragraphs, :actions_taken
+      attr_accessor :failure_reason
+
+      def initialize
+        @actions_taken = []
+        @paragraphs = []
+        @failure_reason = ""
+      end
+
+      def actions_sentence
+        sent = case actions_taken.length
+        when 0 then "No actions taken"
+        when 1 then actions_taken.first
+        when 2 then actions_taken.join(" and ")
+        else
+          actions_taken[0..-2].join(", ") + " and " + actions_taken[-1]
+        end
+        sent.capitalize
+      end
+
+      # Saves a sentence fragment to be displayed in the first sentence of the summary section
```

Comment:
Do you have a good reference for that syntax? I'm happy to use it, but I'm not familiar with it. @kirs made a similar suggestion [previously](https://github.com/Shopify/kubernetes-deploy/pull/24#discussion_r102549877) but just linked to a specific example, so I still don't know what to refer to in general. I don't see the `@example` you're using in https://rdoc.github.io/rdoc/, for instance.